feat: v02 device support

ashleigh-hopkins opened the pull request:
# feat: v02 device support

## Summary

Adds support for V02 devices (Airjet V02, Hydrojet V02, etc.) that use AWS IoT backend instead of Gizwits.

**Dependency:** ⚠️ Builds on #104 - please merge that PR first.

## Background

As discussed in #103, Bestway migrated V02+ devices to AWS IoT Core, leaving these devices unsupported. An alternative integration exists ([M1R4G376/New_bestway_spa](https://github.com/M1R4G376/New_bestway_spa)) but requires Charles Proxy MITM setup and manual credential extraction.

This PR brings V02 support to the main integration, avoiding community fragmentation.

## Implementation Approach

**Adapter Pattern** - AWS IoT API implements the same interface as `BestwayApi`:
- Existing coordinator: 0 lines changed
- Existing entities: only device type filters added
- All new code isolated in `custom_components/bestway/aws_iot/`

Field normalization layer adapts AWS IoT fields (e.g. `water_temperature`) to Gizwits V01 fields (e.g. `Tnow`) so existing entities work unchanged.

## What M1R4G376/New_bestway_spa Provided

Their implementation showed us:
- AWS IoT REST API endpoint (`smarthub-eu.bestwaycorp.com`)
- MD5 authentication signature pattern
- Device shadow GET/POST patterns

## What This PR Adds

**Features not in M1R4G376/New_bestway_spa:**
- AES-256-CBC encrypted commands (v2 API endpoint)
- Real-time WebSocket updates (per-device connections to regional AWS endpoints)
- QR code binding flow (no MITM proxy needed)
- Automatic device discovery via hierarchical API (homes → rooms → devices)
- Multi-device support (one setup finds all devices)
- Regional endpoint selection (EU/US/CN)

**Setup Comparison:**

| M1R4G376/New_bestway_spa | This PR |
|---------|---------|
| Charles Proxy MITM required | No MITM required |
| Manual entry: visitor_id, registration_id, device_id, product_id, device_name | QR code OR visitor_id only |
| One integration per device | One integration discovers all devices |
| REST polling only | WebSocket + polling fallback |

## User Experience

**Setup Flow:**
1. Select backend: V01 (Gizwits) or V02 (AWS IoT)
2. For V02: Select region (EU/US/CN)
3. Scan QR code from app settings OR enter existing visitor_id
4. All devices automatically discovered

**Tested with Airjet V02 (T53NN8):**
- All controls working: power, temperature, heater, filter, bubbles, lock
- Real-time WebSocket updates (<1s latency)

## Production Improvements

- 10s timeout on all API calls (prevents hangs)
- Device discovery caching (reduces API calls from ~1000/day to ~10/day)
- `ConfigEntryAuthFailed` for proper HA reauth flow
- Per-device WebSocket connections to regional AWS endpoints

## Testing

- Manual testing with Airjet V02 (T53NN8) device
- Automated tests added for AWS IoT module (encryption, API client, WebSocket)

## Compatibility

- V01 configurations continue to function
- No migration required
- V01 and V02 can coexist
- Fallback defaults for existing configs

## Closes

#103, #88

## References

#84, #95, #104

---

This PR unifies V01 and V02 support under the main ha-bestway integration.

## Examples 

### V01

<img width="1075" height="936" alt="image" src="https://github.com/user-attachments/assets/8b312a80-c924-4226-bbd1-f21795ed40b4" />

### V02

<img width="1075" height="849" alt="image" src="https://github.com/user-attachments/assets/b256fc6f-ff5a-42e5-81f0-17cf35b9f203" />

ashleigh-hopkins committed:
feat: v02 device support

rayone commented:
Tested with airjet v2, works well.
<img width="2006" height="1647" alt="image" src="https://github.com/user-attachments/assets/f09a601a-9a9b-4b78-a4c8-0b445396e8ca" />

jvogt commented:
Missing controls on ultrafit pump (ultrafit is a new pump which is built into the inflatable hot tub). Happy to add some debug if you can point me in the right direction, or I can share my qr code with you @ashleigh-hopkins. Thanks for the great work!

<img width="1047" height="777" alt="Screenshot 2025-12-09 at 6 17 23 PM" src="https://github.com/user-attachments/assets/5eabb449-0e9a-46d7-9e9e-d95392d3cd0c" />

ashleigh-hopkins committed:
fix: normalize product_series spaces to underscores for ULTRAFIT AIRJET

ashleigh-hopkins commented:
Thanks for reporting! I've just pushed a fix for ULTRAFIT AIRJET devices.

The issue was that the device reports product_series as "ULTRAFIT AIRJET" (with space) but our mapping expected "ULTRAFIT_AIRJET" (with underscore). Now we normalize spaces to underscores during device discovery.

**To test:** Please remove and re-add the integration. The controls (climate, switches, etc.) should now appear for your ULTRAFIT AIRJET device.

Let me know if this resolves it!

ashleigh-hopkins committed:
feat: add V02-specific diagnostic sensors and improvements

- Add V02 diagnostic sensors: WiFi Version, TRD Version, OTA Status
- Hide irrelevant V01 sensors for V02 devices (MCU versions, Protocol Version)
- Replace backend magic strings with BACKEND_GIZWITS/BACKEND_AWS_IOT constants
- Fix WebSocket delta merge to preserve diagnostic fields
- Fix normalization to only include present fields (prevents None overwrite)

V02 devices now show relevant diagnostics from shadow state while
V01 devices continue showing MCU/WiFi versions from device object.

jvogt commented:
@ashleigh-hopkins just tested with my "ULTRAFIT AIRJET" and it's working, thanks and great work!

groovejumper commented:
How would one go about building/installing this? Though I'm a developer, I haven't installed something like this into HA before. I have a new V02 spa with the ultrafit pump (this replaced my older bestway spa from a couple of years ago which worked great with this integration)

ashleigh-hopkins committed:
fix: heater control and display for V01/V02 devices

- Pass through heater_state values (0,1,3,4) unchanged in normalization
- Convert enum value when sending commands (HydrojetHeat.ON=3 → heater_state=1)
- Display HEAT mode for any heat value > 0 (handles all heating states)
- Correctly identify target_reached when heat=4 (not word3 for V02)

Heater state values apply to both V01 and V02:
0=OFF, 1=ON, 3=HEATING, 4=TARGET_REACHED

infernalK commented:
Hi,
It would be great to have this on HA. Is there a way to install/test ?

Stealth100 commented:
How do we install this update please?

cdpuk commented:
Well, I now have water all over my floor thanks to a pump leak during testing and merging of #104, but hopefully that opens up the path to get this finalised. Thanks again for your work on this @ashleigh-hopkins

epheterson commented:
Hey, been running this branch on my Ultrafit Airjet V02 (Fahrenheit) and found an issue:

**Temperature unit flip-flop**

`normalize_aws_state()` defaults `temperature_unit` to `1` (Celsius) when absent. WebSocket partial updates don't always include this field, so the default overwrites the correct `Tunit=0` in the coordinator's dict merge. Result: HA intermittently double-converts Fahrenheit values (79°F → 174°F, 100°F → 212°F).

Fix in `aws_iot/api.py`:
```python
# Before:
normalized["Tunit"] = temperature_unit  # Always include

# After:
if "temperature_unit" in device_state:
    normalized["Tunit"] = temperature_unit
```
Verified on my spa. Happy to create a branch if helpful. 

Thanks for sharing the approach, @ashleigh-hopkins! I've shared a simple python implementation based off of it at https://github.com/epheterson/bestway-spa-v2

ashleigh-hopkins committed:
fix: prevent temperature unit flip-flop on partial WebSocket updates

normalize_aws_state() defaulted temperature_unit to 1 (Celsius) when
absent from WebSocket delta updates, overwriting the real Fahrenheit
setting during coordinator dict merge. Only include Tunit, warning,
and error in normalized output when present in the source data.

Also fix stale test references to non-existent _normalize_state method
and incorrect field names (temp_set/temp_now → Tset/Tnow).

Reported-by: @epheterson

ashleigh-hopkins commented:
## How to install (while this is still a draft PR)

There are two options depending on whether you use HACS:

### Option A: HACS (recommended)

1. Open HACS in your Home Assistant
2. Click the 3-dot menu (top right) → **Custom repositories**
3. Add: `https://github.com/ashleigh-hopkins/ha-bestway` with category **Integration**
4. Search for "Bestway" in HACS and install it
5. **Important:** After install, go to the HACS integration entry → click the 3-dot menu → **Redownload** → select the `feature/dual-backend-support` branch
6. Restart Home Assistant
7. Add the Bestway integration from Settings → Devices & Services → Add Integration

### Option B: Manual install

1. Download the `feature/dual-backend-support` branch as a ZIP from:
   https://github.com/ashleigh-hopkins/ha-bestway/archive/refs/heads/feature/dual-backend-support.zip
2. Extract and copy the `custom_components/bestway` folder into your HA `config/custom_components/` directory (overwrite existing if upgrading)
3. Restart Home Assistant
4. Add the Bestway integration from Settings → Devices & Services → Add Integration

### Setup (V02 devices)

- Select "V02 (AWS IoT)" as the backend
- Choose your region (EU/US/CN)
- Either scan the QR code from the Lay-Z-Spa app (Settings → QR Code) or enter your visitor_id
- All devices are discovered automatically

**Note:** If you already have the V01 (Gizwits) integration, it will continue to work. V01 and V02 can coexist.

### Recent fixes in this update

- Rebased onto latest main (includes gizwits websocket from #104)
- Fixed temperature unit flip-flop for Fahrenheit users (thanks @epheterson)
- Fixed all linting and type checking issues
- All 67 tests passing

ashleigh-hopkins committed:
fix: type safety, linting, and stale test cleanup

- Replace Any with proper types (ClientSession, GizwitsWebSocket,
  AwsIotWebSocket, BestwayApi | AwsIotApi)
- Fix mypy errors across aws_iot/api.py, websocket.py, __init__.py
- Fix ruff lint (unused variable, formatting)
- Update stale tests to match actual API (normalize_aws_state,
  correct field names, _do_get/_do_post patching)
- Add missing airjet_spa_set_heat and pool filter stubs to AwsIotApi
- Add node_modules to .gitignore

Stealth100 commented:
I have managed to get it showing v02 setup option in home assistant by copying all files individually in to specific locations in file editor. I select qr code option but don't bring up any way to scan in? Where would I find visitor id as the other option please?

Stealth100 commented:
> ## How to install (while this is still a draft PR)
> There are two options depending on whether you use HACS:
> 
> ### Option A: HACS (recommended)
> 1. Open HACS in your Home Assistant
> 2. Click the 3-dot menu (top right) → **Custom repositories**
> 3. Add: `https://github.com/ashleigh-hopkins/ha-bestway` with category **Integration**
> 4. Search for "Bestway" in HACS and install it
> 5. **Important:** After install, go to the HACS integration entry → click the 3-dot menu → **Redownload** → select the `feature/dual-backend-support` branch
> 6. Restart Home Assistant
> 7. Add the Bestway integration from Settings → Devices & Services → Add Integration
> 
> ### Option B: Manual install
> 1. Download the `feature/dual-backend-support` branch as a ZIP from:
>    https://github.com/ashleigh-hopkins/ha-bestway/archive/refs/heads/feature/dual-backend-support.zip
> 2. Extract and copy the `custom_components/bestway` folder into your HA `config/custom_components/` directory (overwrite existing if upgrading)
> 3. Restart Home Assistant
> 4. Add the Bestway integration from Settings → Devices & Services → Add Integration
> 
> ### Setup (V02 devices)
> * Select "V02 (AWS IoT)" as the backend
> * Choose your region (EU/US/CN)
> * Either scan the QR code from the Lay-Z-Spa app (Settings → QR Code) or enter your visitor_id
> * All devices are discovered automatically
> 
> **Note:** If you already have the V01 (Gizwits) integration, it will continue to work. V01 and V02 can coexist.
> 
> ### Recent fixes in this update
> * Rebased onto latest main (includes gizwits websocket from [feat: gizwits websocket implementation #104](https://github.com/cdpuk/ha-bestway/pull/104))
> * Fixed temperature unit flip-flop for Fahrenheit users (thanks @epheterson)
> * Fixed all linting and type checking issues
> * All 67 tests passing

I have used another device to share to using qr code from bestway connect app, can't scan any qr from within ha at device register page for v102 so I used the guest id from the device I just shared but comes up no devices.

ashleigh-hopkins commented:
You can scan the QR code using another device, or take a screenshot, crop it and upload it to an online qr code parser. The QR code contains a string which you can paste into the setup flow.

Stealth100 commented:
> You can scan the QR code using another device, or take a screenshot, crop it and upload it to an online qr code parser. The QR code contains a string which you can paste into the setup flow.

Ok so paste the string in to qr part on ha?

Stealth100 commented:
> You can scan the QR code using another device, or take a screenshot, crop it and upload it to an online qr code parser. The QR code contains a string which you can paste into the setup flow.

Brilliant that got ot sort for the confusion

ashleigh-hopkins commented:
Also, for the visitor_id this is only useful if you've MITM'd your app and collected your visitor id OR after you've used a QR code in this integration, it's made a new account for you and recover it from the logs, etc. I provided both options to support the existing flow from the other v02 standalone integration plus it was useful during testing!

epheterson commented:
One more thought — would it make sense to expose `current_temperature` and `target_temperature` as dedicated sensor entities (in addition to the climate entity)?

Right now temperature is an attribute on the climate entity, so HA's recorder only captures it when the HVAC state or attributes change. A `sensor.spa_current_temperature` whose state IS the temperature value would integrate more naturally with HA's statistics, history graphs, and energy dashboard. It would also make it easy to set up automations that trigger on temperature thresholds.

Most thermostat integrations (Ecobee, Nest, etc.) expose separate temperature sensors alongside the climate entity for exactly this reason.

jvogt commented:
> Most thermostat integrations (Ecobee, Nest, etc.) expose separate temperature sensors alongside the climate entity for exactly this reason.

That alone gets my vote yes on this idea

james-rival commented:
Many thanks for doing this. I also got a v02 pump as a warranty replacement which of course stopped me using a number of HA automations I'd set up.

Unfortunately I'm having a challenging installing this at step 5:

Important: After install, go to the HACS integration entry → click the 3-dot menu → Redownload → select the feature/dual-backend-support branch

I'm not getting any options when I re-download, it just says:
Commit 69e7136 will be downloaded

I do get multiple options for re-download on the original cpduk integration.

epheterson commented:
I still think the sensors are probably worth adding, in my implementation it's working well and makes it easy to chart and look at data over time.

Docrob9 commented:
Awesome integration & light years easier than using the Charles approach - thank you so much!
